File: app/what_changed.py
"""Deterministic, dated league movers over consecutive team-game windows."""

from __future__ import annotations

from collections import defaultdict
from datetime import date, timedelta
from math import isfinite
from typing import Any, Literal

ComparisonPeriod = Literal["four_games", "week"]
SeasonPhase = Literal["Regular Season", "Playoffs"]


class WhatChangedUnavailable(RuntimeError):
    """The bounded source read cannot safely produce a complete league ranking."""


METRICS = {
    "availability": [("min", "Minutes / appearance"), ("pf", "Fouls / appearance")],
    "offense": [
        ("pts", "Points"),
        ("ast", "Assists"),
        ("fg3m", "Made threes"),
        ("oreb", "Offensive rebounds"),
        ("tov", "Turnovers"),
        ("fg_pct", "FG%"),
        ("ft_pct", "FT%"),
        ("ts_pct", "TS%"),
    ],
    "defense": [("dreb", "Defensive rebounds"), ("stl", "Steals"), ("blk", "Blocks")],
}
COUNTING = {
    "min",
    "pf",
    "pts",
    "reb",
    "ast",
    "stl",
    "blk",
    "tov",
    "fg3m",
    "oreb",
    "dreb",
    "fgm",
    "fga",
    "ftm",
    "fta",
    "fg3a",
}
PRODUCTION = ("pts", "reb", "ast", "stl", "blk", "tov")


def _date(value: Any) -> date:
    return date.fromisoformat(str(value)[:10])


def _number(value: Any) -> float | None:
    try:
        number = float(value)
        return number if isfinite(number) else None
    except (ValueError, TypeError):
        return None


def _rounded(value: float | None) -> float | None:
    return round(value, 2) if value is not None else None


def _aggregate(rows: list[dict]) -> dict:
    played = [row for row in rows if (_number(row.get("min")) or 0) > 0]
    totals: dict[str, float | None] = {}
    for key in COUNTING:
        observations = [_number(row.get(key)) for row in played]
        totals[key] = (
            sum(value for value in observations if value is not None)
            if observations and all(value is not None for value in observations)
            else None
        )
    values = {
        key: total / len(played) if total is not None and played else None
        for key, total in totals.items()
    }
    for key, made, attempted in [("fg_pct", "fgm", "fga"), ("ft_pct", "ftm", "fta")]:
        numerator, denominator = totals[made], totals[attempted]
        values[key] = (
            100 * numerator / denominator
            if numerator is not None and denominator
            else None
        )
    pts, fga, fta = totals["pts"], totals["fga"], totals["fta"]
    attempts = (fga + 0.44 * fta) if fga is not None and fta is not None else None
    values["ts_pct"] = (
        100 * pts / (2 * attempts) if pts is not None and attempts else None
    )
    production = (
        sum(float(values[key] or 0) * (-1 if key == "tov" else 1) for key in PRODUCTION)
        if all(values[key] is not None for key in PRODUCTION)
        else None
    )
    return {
        "games_played": len(played),
        "values": values,
        "totals": totals,
        "production": production,
    }


def _metric(key: str, label: str, current: dict, previous: dict) -> dict:
    value, prior = current["values"].get(key), previous["values"].get(key)
    return {
        "key": key,
        "label": label,
        "unit": "percent" if key.endswith("pct") else "count",
        "current": _rounded(value),
        "previous": _rounded(prior),
        "delta": _rounded(value - prior)
        if value is not None and prior is not None
        else None,
        "current_total": _rounded(current["totals"].get(key)),
        "previous_total": _rounded(previous["totals"].get(key)),
        "lower_is_better": key in {"tov", "pf"},
    }


def _injuries(
    player_id: int,
    team: str,
    games: list[dict],
    reports: list[dict] | None,
    as_of: date,
) -> dict:
    if reports is None:
        return {
            "available": False,
            "games_covered": 0,
            "out_reports": None,
            "reports": [],
        }
    game_dates = {game["game_date"] for game in games}
    latest: dict[str, dict] = {}
    for row in reports:
        if row.get("player_id") != player_id or row.get("team_abbr") != team:
            continue
        game_date = str(row.get("game_date", ""))[:10]
        report_date = str(row.get("report_date", ""))[:10]
        # A report is evidence of a reported status, never proof of a missed game.
        if (
            game_date not in game_dates
            or not report_date
            or report_date > min(game_date, as_of.isoformat())
        ):
            continue
        stamp = str(row.get("report_timestamp_utc") or report_date)
        if stamp[:10] > as_of.isoformat():
            continue
        if game_date not in latest or stamp > str(
            latest[game_date].get("report_timestamp_utc")
            or latest[game_date].get("report_date")
        ):
            latest[game_date] = row
    selected = [latest[key] for key in sorted(latest)]
    return {
        "available": True,
        "games_covered": len(selected),
        "out_reports": sum(row.get("injury_status") == "Out" for row in selected)
        if selected
        else None,
        "reports": [
            {
                key: row.get(key)
                for key in [
                    "game_date",
                    "report_date",
                    "report_timestamp_utc",
                    "injury_status",
                    "reason",
                    "source_url",
                ]
            }
            for row in selected
        ],
    }


def _window(
    games: list[dict],
    rows: list[dict],
    player_id: int,
    team: str,
    reports: list[dict] | None,
    as_of: date,
) -> dict:
    by_game = {row["game_id"]: row for row in rows if row.get("team_abbr") == team}
    selected = [
        by_game[game["game_id"]] for game in games if game["game_id"] in by_game
    ]
    result = _aggregate(selected)
    result.update(
        {
            "team_games": len(games),
            "start_date": min((game["game_date"] for game in games), default=None),
            "end_date": max((game["game_date"] for game in games), default=None),
            "no_recorded_appearance": len(games) - result["games_played"],
            "dnp_cd": None,
            "injuries": _injuries(player_id, team, games, reports, as_of),
            "games": [
                {
                    **game,
                    "played": (
                        _number(by_game.get(game["game_id"], {}).get("min")) or 0
                    )
                    > 0,
                    "stats": {
                        key: _number(by_game.get(game["game_id"], {}).get(key))
                        for key in [
                            "min",
                            "pts",
                            "reb",
                            "ast",
                            "stl",
                            "blk",
                            "tov",
                            "pf",
                        ]
                    },
                }
                for game in sorted(
                    games, key=lambda game: (game["game_date"], game["game_id"])
                )
            ],
        }
    )
    return result


def build_what_changed(
    game_rows: list[dict],
    injury_rows: list[dict] | None,
    *,
    period: ComparisonPeriod = "four_games",
    season_type: SeasonPhase = "Regular Season",
    as_of: str | None = None,
) -> dict[str, Any]:
    if period not in {"four_games", "week"} or season_type not in {
        "Regular Season",
        "Playoffs",
    }:
        raise ValueError("Unsupported comparison window or season type")
    requested = _date(as_of) if as_of else None
    # Corrections are deduplicated before either game counting or aggregation.
    deduped: dict[tuple[int, str], dict] = {}
    for row in sorted(game_rows, key=lambda row: str(row.get("ingested_at_utc") or "")):
        if (
            row.get("season_type") != season_type
            or not row.get("game_id")
            or not row.get("player_id")
        ):
            continue
        if requested and _date(row["game_date"]) > requested:
            continue
        deduped[(int(row["player_id"]), str(row["game_id"]))] = {
            **row,
            "game_id": str(row["game_id"]),
            "game_date": _date(row["game_date"]).isoformat(),
        }
    rows = list(deduped.values())
    data_through = max((_date(row["game_date"]) for row in rows), default=None)
    payload: dict[str, Any] = {
        "period": period,
        "season_type": season_type,
        "data_through": data_through.isoformat() if data_through else None,
        "requested_as_of": as_of,
        "window_basis": "team_games",
        "items": [],
        "injury_source_available": injury_rows is not None,
        "dnp_cd_available": False,
        "ranking": {
            "formula": "PTS + REB + AST + STL + BLK − TOV per recorded appearance",
            "top": "Highest recent production index",
            "surging": "Largest positive change from the previous window",
            "minimum_games": 3 if period == "four_games" else 2,
            "minimum_minutes_per_appearance": 10,
            "note": "A transparent box-score proxy, not your fantasy league's scoring or an all-around player rating.",
        },
    }
    if data_through is None:
        payload["state"] = "empty"
        return payload
    anchor = min(requested, data_through) if requested else data_through
    if period == "week":
        # Compare two complete Monday–Sunday weeks anchored to available data.
        end = (
            anchor
            if anchor.weekday() == 6
            else anchor - timedelta(days=anchor.weekday() + 1)
        )
        current_start, previous_start = (
            end - timedelta(days=6),
            end - timedelta(days=13),
        )
        payload["calendar_windows"] = {
            "current": {"start": current_start.isoformat(), "end": end.isoformat()},
            "previous": {
                "start": previous_start.isoformat(),
                "end": (current_start - timedelta(days=1)).isoformat(),
            },
        }
        anchor = end
    payload["as_of"] = anchor.isoformat()
    rows = [row for row in rows if _date(row["game_date"]) <= anchor]
    teams: dict[str, dict[str, dict]] = defaultdict(dict)
    players: dict[int, list[dict]] = defaultdict(list)
    for row in rows:
        team, opponent = row.get("team_abbr"), row.get("opponent_abbr")
        if not team:
            continue
        players[int(row["player_id"])].append(row)
        for team_key, other in [(team, opponent), (opponent, team)]:
            if team_key:
                teams[team_key][row["game_id"]] = {
                    "game_id": row["game_id"],
                    "game_date": row["game_date"],
                    "opponent": other,
                }
    for player_id, history in players.items():
        latest = max(history, key=lambda row: (row["game_date"], row["game_id"]))
        team = latest["team_abbr"]
        team_games = sorted(
            teams[team].values(),
            key=lambda game: (game["game_date"], game["game_id"]),
            reverse=True,
        )
        if not team_games or _date(team_games[0]["game_date"]) < anchor - timedelta(
            days=6
        ):
            continue  # Do not present eliminated/inactive teams as current movers.
        if period == "four_games":
            current_games, previous_games = team_games[:4], team_games[4:8]
        else:
            current_games = [
                game
                for game in team_games
                if current_start <= _date(game["game_date"]) <= anchor
            ]
            previous_games = [
                game
                for game in team_games
                if previous_start <= _date(game["game_date"]) < current_start
            ]
        current = _window(current_games, history, player_id, team, injury_rows, anchor)
        previous = _window(
            previous_games, history, player_id, team, injury_rows, anchor
        )
        minimum = payload["ranking"]["minimum_games"]
        top_eligible = (
            current["games_played"] >= minimum
            and (current["values"]["min"] or 0) >= 10
            and current["production"] is not None
        )
        if period == "four_games":
            top_eligible = top_eligible and current["team_games"] == 4
        comparison_start = previous["start_date"] or current["start_date"]
        team_changed = (
            any(
                row["team_abbr"] != team and row["game_date"] >= comparison_start
                for row in history
            )
            if comparison_start
            else False
        )
        surge_eligible = (
            top_eligible
            and previous["games_played"] >= minimum
            and previous["production"] is not None
            and not team_changed
        )
        if period == "four_games":
            surge_eligible = surge_eligible and previous["team_games"] == 4
        delta = (
            current["production"] - previous["production"]
            if current["production"] is not None and previous["production"] is not None
            else None
        )
        clusters = [
            {
                "key": key,
                "label": {
                    "availability": "Playing availability",
                    "offense": "Offense",
                    "defense": "Defense",
                }[key],
                "metrics": [
                    _metric(stat, label, current, previous) for stat, label in metrics
                ],
            }
            for key, metrics in METRICS.items()
        ]
        payload["items"].append(
            {
                "player_id": player_id,
                "player_name": latest["player_name"],
                "team_abbr": team,
                "current": current,
                "previous": previous,
                "clusters": clusters,
                "top_score": _rounded(current["production"]) if top_eligible else None,
                "surge_score": _rounded(delta)
                if surge_eligible and delta is not None and delta > 0
                else None,
                "production_delta": _rounded(delta),
                "team_changed": team_changed,
                "sample_note": "Team changed within the comparison; surge ranking excluded."
                if team_changed
                else "Limited comparison sample."
                if not surge_eligible
                else "Small descriptive sample; not a forecast.",
            }
        )
    payload["items"].sort(
        key=lambda item: (
            -(item["top_score"] if item["top_score"] is not None else float("-inf")),
            item["player_id"],
        )
    )
    for score, rank_key in [("top_score", "top_rank"), ("surge_score", "surge_rank")]:
        eligible = sorted(
            (item for item in payload["items"] if item[score] is not None),
            key=lambda item: (-item[score], item["player_id"]),
        )
        for rank, item in enumerate(eligible, 1):
            item[rank_key] = rank
    payload["state"] = "ready" if payload["items"] else "empty"
    return payload

File: app/static/what_changed.js
import { seasonFetch } from "./season.js";
const escape = (value) => String(value ?? "").replace(/[&<>"']/g, (char) => ({"&":"&amp;", "<":"&lt;", ">":"&gt;", '"':"&quot;", "'":"&#39;"}[char]));
export const number = (value, signed = false) => value == null || !Number.isFinite(Number(value)) ? "—" : `${signed && Number(value) > 0 ? "+" : ""}${Number(value).toFixed(1)}`;
export function selectMovers(items, view, search = "") {
  const score = view === "surging" ? "surge_score" : "top_score";
  const query = search.trim().toLocaleLowerCase();
  return items.filter((item) => item[score] != null && (!query || `${item.player_name} ${item.team_abbr}`.toLocaleLowerCase().includes(query)))
    .sort((a, b) => b[score] - a[score] || a.player_id - b.player_id);
}
function dates(window) { return window?.start_date ? `${window.start_date} → ${window.end_date}` : "No comparison games"; }
function safeSource(url) {
  try { const parsed = new URL(url); return parsed.protocol === "https:" ? escape(parsed.href) : ""; } catch { return ""; }
}
function metricTable(cluster) {
  return `<table><thead><tr><th scope="col">Metric</th><th scope="col">Recent</th><th scope="col">Prior</th><th scope="col">Change</th></tr></thead><tbody>${cluster.metrics.map((metric) => {
    const delta = metric.delta;
    const favorable = delta == null || delta === 0 ? "" : (metric.lower_is_better ? delta < 0 : delta > 0) ? "delta-positive" : "delta-negative";
    return `<tr><th scope="row">${escape(metric.label)}</th><td>${number(metric.current)}${metric.unit === "percent" ? "%" : ""}${metric.current_total != null ? `<small>${number(metric.current_total)} total</small>` : ""}</td><td>${number(metric.previous)}${metric.unit === "percent" ? "%" : ""}${metric.previous_total != null ? `<small>${number(metric.previous_total)} total</small>` : ""}</td><td class="${favorable}">${number(delta, true)}${metric.unit === "percent" && delta != null ? " pp" : ""}</td></tr>`;
  }).join("")}</tbody></table>`;
}
function availability(item) {
  const c = item.current, p = item.previous;
  const reportList = c.injuries.reports.map((report) => {
    const url = safeSource(report.source_url);
    const text = `${report.game_date}: ${report.injury_status} — ${report.reason || "No reason supplied"}`;
    return `<li>${url ? `<a href="${url}" target="_blank" rel="noopener noreferrer">${escape(text)}</a>` : escape(text)} <span>(report ${escape(report.report_date)})</span></li>`;
  }).join("");
  return `<p class="metric-note">Recorded appearances: <strong>${c.games_played}/${c.team_games}</strong> recent · ${p.games_played}/${p.team_games} prior.<br>No recorded appearance: ${c.no_recorded_appearance} recent · ${p.no_recorded_appearance} prior. Reasons are unconfirmed.</p>
  <p class="metric-note">Injury-report coverage: ${c.injuries.games_covered}/${c.team_games} recent · ${p.injuries.games_covered}/${p.team_games} prior. ${c.injuries.available ? "Missing reports do not establish health." : "Historical injury reports are unavailable."}<br>Games with an Out report: ${number(c.injuries.out_reports)} recent · ${number(p.injuries.out_reports)} prior. These are reported statuses, not confirmed missed games.<br>DNP-CD: not available from this source.</p>${reportList ? `<ul class="mover-injuries">${reportList}</ul>` : ""}`;
}
export function renderMover(item, view) {
  const score = view === "surging" ? item.surge_score : item.top_score;
  const rank = view === "surging" ? item.surge_rank : item.top_rank;
  return `<article class="mover-card" data-player-id="${Number(item.player_id)}">
    <div class="mover-heading"><div><h2><span class="mover-rank">${rank}.</span><a href="/players/${Number(item.player_id)}">${escape(item.player_name)}</a></h2><p class="meta">${escape(item.team_abbr)} · ${item.current.games_played} recent appearances</p></div><div class="mover-score">${number(score, view === "surging")}<small>${view === "surging" ? "production increase / appearance" : "production index / appearance"}</small></div></div>
    <p class="mover-sample">Recent ${dates(item.current)} · Prior ${dates(item.previous)}<br>${escape(item.sample_note)}</p>
    <div class="mover-clusters">${item.clusters.map((cluster) => {
      const headline = cluster.metrics[0];
      return `<details class="mover-cluster"><summary>${escape(cluster.label)}<span>${escape(headline.label)}: ${number(headline.current)} · ${number(headline.delta, true)}</span></summary>${metricTable(cluster)}${cluster.key === "availability" ? availability(item) : `<p class="metric-note">Counting stats are per appearance; totals appear underneath.${cluster.key === "defense" ? " Box-score events do not measure all defensive impact." : " Shooting changes are percentage points."}</p>`}</details>`;
    }).join("")}</div>
    <details class="mover-games"><summary>Inspect the games behind this comparison</summary><div class="table-scroll"><table><thead><tr><th scope="col">Window / date</th><th scope="col">Opponent</th><th scope="col">MIN</th><th scope="col">PTS</th><th scope="col">REB</th><th scope="col">AST</th><th scope="col">STL</th><th scope="col">BLK</th><th scope="col">TOV</th></tr></thead><tbody>${[["Recent", item.current], ["Prior", item.previous]].flatMap(([label, window]) => window.games.map((game) => `<tr><th scope="row">${label} ${escape(game.game_date)}${game.played ? "" : "<br>No recorded appearance"}</th><td>${escape(game.opponent)}</td>${["min", "pts", "reb", "ast", "stl", "blk", "tov"].map((key) => `<td>${number(game.stats[key])}</td>`).join("")}</tr>`)).join("")}</tbody></table></div></details>
  </article>`;
}

// A point requires a comparable baseline; high-production players can still
// appear in the ranked list when their previous window is insufficient.
export function opportunityPoints(items, minimum = 3, fullWindow = true) {
  return items.flatMap((item) => {
    const recent = item.current?.values?.min, prior = item.previous?.values?.min;
    const delta = item.production_delta;
    if (recent == null || prior == null || delta == null || item.team_changed
      || item.previous.games_played < minimum || (fullWindow && item.previous.team_games < 4)
      || ![recent, prior, delta].every(Number.isFinite)) return [];
    return [{item, x: recent - prior, y: delta}];
  });
}
export function chartAxis(values) {
  const low = Math.min(0, ...values), high = Math.max(0, ...values);
  const rough = Math.max(high - low, 2) / 5;
  const power = 10 ** Math.floor(Math.log10(rough));
  const step = [1, 2, 5, 10].map(value => value * power).find(value => value >= rough);
  const min = Math.floor(low / step) * step - step, max = Math.ceil(high / step) * step + step;
  const ticks = [];
  for (let i = 0; i <= Math.round((max - min) / step); i++) ticks.push(Number((min + i * step).toFixed(6)));
  return {min, max, ticks};
}
export function renderOpportunityMap(points, selectedId) {
  if (!points.length) return '<p class="empty-state">No comparable prior windows are available to plot. Player details remain available in the ranked list.</p>';
  const xa = chartAxis(points.map(point => point.x)), ya = chartAxis(points.map(point => point.y));
  const x = value => 75 + (value - xa.min) / (xa.max - xa.min) * 590;
  const y = value => 395 - (value - ya.min) / (ya.max - ya.min) * 335;
  const selected = points.find(point => point.item.player_id === selectedId);
  // The selected point is drawn last so it stays visible when coordinates overlap.
  const ordered = points.filter(point => point !== selected).concat(selected ? [selected] : []);
  return `<div class="opportunity-scroll"><svg viewBox="0 0 740 465" role="group" aria-label="Opportunity map: minutes change versus production change">
    <text x="18" y="22" class="axis-title">PRODUCTION CHANGE / APPEARANCE</text>
    ${ya.ticks.map(value => `<line x1="75" x2="665" y1="${y(value)}" y2="${y(value)}" class="${value === 0 ? 'zero-line' : 'grid-line'}"/><text x="62" y="${y(value)+4}" text-anchor="end">${number(value,true)}</text>`).join('')}
    ${xa.ticks.map(value => `<line x1="${x(value)}" x2="${x(value)}" y1="60" y2="395" class="${value === 0 ? 'zero-line' : 'grid-line'}"/><text x="${x(value)}" y="420" text-anchor="middle">${number(value,true)}</text>`).join('')}
    <text x="370" y="453" text-anchor="middle" class="axis-title">MINUTES CHANGE / APPEARANCE</text>
    ${ordered.map(point => `<g role="button" tabindex="0" data-select-player="${Number(point.item.player_id)}" aria-pressed="${point.item.player_id === selectedId}" aria-label="Select ${escape(point.item.player_name)}: minutes ${number(point.x,true)}, production ${number(point.y,true)}"><title>${escape(point.item.player_name)} · ${escape(point.item.team_abbr)}: minutes ${number(point.x,true)}, production ${number(point.y,true)}</title><circle class="point-hit" cx="${x(point.x)}" cy="${y(point.y)}" r="13"/><circle class="point" cx="${x(point.x)}" cy="${y(point.y)}" r="${point.item.player_id === selectedId ? 7 : 4}"/></g>`).join('')}
    ${selected ? `<text class="point-label" x="${x(selected.x)}" y="${y(selected.y)-18}" text-anchor="${x(selected.x) > 540 ? 'end' : x(selected.x) < 180 ? 'start' : 'middle'}">${escape(selected.item.player_name)}</text>` : ''}
  </svg></div>`;
}

export function initWhatChanged(doc = document, fetcher = seasonFetch) {
  const form = doc.querySelector("#changed-controls");
  if (!form) return;
  let payload = null, view = "top", shown = 20, requestId = 0, selectedId = null;
  const list = doc.querySelector("#changed-list"), status = doc.querySelector("#changed-status"), more = doc.querySelector("#changed-more"), search = doc.querySelector("#changed-search");
  const chart = doc.querySelector("#changed-chart"), detail = doc.querySelector("#changed-detail");
  function bindSelections(root) {
    root.querySelectorAll("[data-select-player]").forEach(node => {
      const select = () => {
        selectedId = Number(node.dataset.selectPlayer);
        const inChart = Boolean(node.closest("#changed-chart"));
        render();
        const target = (inChart ? chart : list).querySelector(`[data-select-player="${selectedId}"]`);
        target?.focus({preventScroll:true});
      };
      node.addEventListener("click", select);
      if (node.tagName.toLowerCase() === "g") node.addEventListener("keydown", event => {
        if (event.key === "Enter" || event.key === " ") { event.preventDefault(); select(); }
      });
    });
  }
  function render() {
    if (!payload) return;
    const items = selectMovers(payload.items, view, search.value);
    const excluded = payload.items.filter((item) => item.top_score == null).length;
    doc.querySelector("#changed-ranking").textContent = `${view === "surging" ? payload.ranking.surging : payload.ranking.top}. ${payload.ranking.formula}.`;
    status.textContent = items.length ? `${items.length} matching players. ${excluded} players excluded from top rankings for limited samples or minutes.` : payload.state === "empty" ? "No games are available for this season phase and date." : view === "surging" ? "No players meet the positive-surge and sample requirements for these filters." : "No players meet the sample and minutes requirements for these filters.";
    if (!items.some(item => item.player_id === selectedId)) selectedId = items[0]?.player_id ?? null;
    const points = opportunityPoints(items, payload.ranking.minimum_games, payload.period === "four_games");
    chart.innerHTML = items.length ? renderOpportunityMap(points, selectedId) : "";
    doc.querySelector("#changed-chart-note").textContent = items.length ? `${points.length} of ${items.length} matching players have comparable baselines. Select a point or ranked player. Overlapping points remain individually selectable below. More minutes and more production appear above and right of zero; this does not establish causation.` : "";
    list.innerHTML = items.slice(0, shown).map(item => `<button type="button" class="mover-select" data-select-player="${Number(item.player_id)}" aria-pressed="${item.player_id === selectedId}"><span><strong>${escape(item.player_name)}</strong><small>${escape(item.team_abbr)} · ${item.current.games_played} appearances</small></span><span class="mover-select-score">${number(view === "surging" ? item.surge_score : item.top_score, view === "surging")}<small>${view === "surging" ? 'production change' : 'production index'}</small></span></button>`).join("");
    const selected = items.find(item => item.player_id === selectedId);
    detail.innerHTML = selected ? `<p class="eyebrow">Selected player</p>${renderMover(selected, view)}` : "";
    detail.querySelectorAll('.mover-cluster').forEach(node => {node.open = true;});
    bindSelections(chart); bindSelections(list);
    more.hidden = items.length <= shown;
  }
  async function load() {
    const id = ++requestId;
    payload = null;
    list.innerHTML = "";
    chart.innerHTML = ""; detail.innerHTML = "";
    doc.querySelector("#changed-chart-note").textContent = "";
    list.setAttribute("aria-busy", "true");
    more.hidden = true;
    status.textContent = "Loading league movers…";
    doc.querySelector("#changed-dates").textContent = "";
    doc.querySelector("#changed-ranking").textContent = "";
    const params = new URLSearchParams();
    for (const [key, value] of new FormData(form)) if (value) params.set(key, value);
    try {
      const response = await fetcher(`/api/what-changed?${params}`);
      if (!response.ok) throw new Error("unavailable");
      const result = await response.json();
      if (id !== requestId) return;
      payload = result; shown = 20;
      const weeks = result.calendar_windows;
      doc.querySelector("#changed-dates").textContent = weeks
        ? `Recent week ${weeks.current.start} → ${weeks.current.end} · Previous week ${weeks.previous.start} → ${weeks.previous.end}. Source games through ${result.data_through}.`
        : `Latest games through ${result.data_through || "—"}. Each team’s last four games are compared with its previous four; selected player details show exact dates.`;
      render();
    } catch {
      if (id === requestId) status.textContent = "Unable to load a complete league comparison. Try Update again; no partial ranking is shown.";
    } finally { if (id === requestId) list.setAttribute("aria-busy", "false"); }
  }
  form.addEventListener("submit", (event) => { event.preventDefault(); load(); });
  search.addEventListener("input", () => { shown = 20; render(); });
  doc.querySelectorAll("[data-mover-view]").forEach((button) => button.addEventListener("click", () => {
    view = button.dataset.moverView; shown = 20;
    doc.querySelectorAll("[data-mover-view]").forEach((node) => { const selected = node === button; node.classList.toggle("is-active", selected); node.setAttribute("aria-pressed", String(selected)); });
    render();
  }));
  more.addEventListener("click", () => { shown += 20; render(); });
  load();
}
if (typeof document !== "undefined") initWhatChanged();
